packages/core/content-manager/server/src/history/services/history.ts

```diff
@@ -236,7 +236,25 @@ const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
         structuredClone(dataWithoutAddedAttributes)
       );
 
-      const data = omit(['id', ...Object.keys(schemaDiff.removed)], dataWithoutMissingRelations);
+      /**
+       * Force Strapi to re-create the components when restoring by removing the previous ID.
+       * This is necessary because the previous component row in the database may have been deleted,
+       * which would cause an error when referencing its ID.
+       */
+      const dataWithoutComponentIds = await traverseEntity(
+        ({ parent }, { remove }) => {
+          if (['dynamiczone', 'component'].includes(parent?.attribute?.type ?? '')) {
+            remove('id');
+          }
+        },
+        {
+          schema: strapi.getModel(version.contentType),
+          getModel: strapi.getModel.bind(strapi),
+        },
+        dataWithoutMissingRelations
+      );
+
+      const data = omit(['id', ...Object.keys(schemaDiff.removed)], dataWithoutComponentIds);
```

Comment:
I was wondering if it would be easier to not select the id field from compos and dz's from the beggining, but that would not work for history versions created before this fix.

Reply:
You're right! A big plus of your approach is not having to apply that logic each time history is fetched.

It won't work for history versions from 5.0.0 to 5.0.2, but I don't think it's a big deal, history versions get auto-deleted after the retention period anyway